Add backward incompatible changes for 2201.4.0 (#21)
praneesha/ballerina-dev-website: downloads/swan-lake-release-notes/swan-lake-2201.4.0/RELEASE_NOTE.md

To view other bug fixes, see the [GitHub milestone for Swan Lake 2201.4.0](https://github.com/ballerina-platform/ballerina-lang/issues?q=is%3Aissue+label%3ATeam%2FCompilerFE+milestone%3A2201.4.0+is%3Aclosed+label%3AType%2FBug).

#### Backward incompatible changes

- Fixed a bug that incorrectly resolved the result type of a query action that would complete normally to `error?` instead of `()`. Now, the result type includes `error` only in instances where an error can be generated during the execution of the query pipeline (`from` clause/`join` clause).

```ballerina
public function main() returns error? {
    from int i in 1 ... 3 // now valid
    do {
    };
}

function iterateStream(stream<int, error?> numberStream) returns error? {
    check from int num in numberStream
        do {
        };
}
```

> - Fixed a bug that incorrectly propagated errors returned from a `do` clause of a query action to the result of the query action. Now, if the execution of a statement within a `do` clause fails with an error, it will be propagated to the nearest enclosing failure-handling statement.

Reviewer: ```suggestion
- Fixed a bug that incorrectly propagated errors returned in a `do` clause of a query action to the result of the query action. Now, if the execution of a statement within a `do` clause fails with an error, it will be propagated to the nearest enclosing failure-handling statement.
```

```ballerina

> public function main() {
>     error? queryActResult = from int i in 1 ... 3
>         do {
>             check validateAndGetError(); // error: invalid usage of the 'check' expression operator; 
>                                          // no matching error return type(s) in the enclosing invokable
>         };
> }
> 
> function validateAndGetError() returns error? {
>     return error("Invalid error");
> }

Reviewer: How about

```suggestion
public function main() returns error? {
    from int i in 1 ... 3
        do {
            check validateAndGetError(i); // now valid
        };
}

function validateAndGetError(int i) returns error? {
    return i > 0 ? () : error("Invalid valid");
}
```

Author: The example should say when we have a `check` exp inside do-clause, it will be propagated outside not to the query result even when the expected type for query-action includes an error. Any other suggestion?

To view other bug fixes, see the [GitHub milestone for Swan Lake 2201.4.0](https://github.com/ballerina-platform/ballerina-lang/issues?q=is%3Aissue+label%3ATeam%2FCompilerFE+milestone%3A2201.4.0+is%3Aclosed+label%3AType%2FBug).

#### Backward incompatible changes

> - Fixed a bug that incorrectly resolved the result type of a query action that would complete normally to `error?` instead of `()`.
>   Now, resolution of the result type of a query action has been improved so that it is now set to `error` only in instances where an error can be generated during the execution of the query pipeline (`from-clause`/`join-clause`).

Reviewer: ```suggestion
- Fixed a bug that incorrectly resolved the result type of a query action that would complete normally to `error?` instead of `()`. Now, the result type includes `error` only in instances where an error can be generated during the execution of the query pipeline (`from` clause/`join` clause).
```

Reviewer: I don't think we need the separate lines (with the first line being like the title.).

```ballerina

> public function main() returns error? {
>     from int i in 1 ... 3
>     do {
>     };
> 
>     check from int i in 1 ... 3 // warning: invalid usage of the 'check' expression operator: no expression type is equivalent to error type
>         do {
>         };

Reviewer: Should we just say

```suggestion
public function main() returns error? {
    from int i in 1 ... 3 // now valid
    do {
    };
```

Reviewer: I think we shouldn't add examples of wrong code if not really required to explain the point.

}

function iterateStream(stream<int, error?> numberStream) returns error? {
    check from int num in numberStream
        do {
        };
}
```

> - Fixed a bug of propagating errors thrown from the `do-clause` in a query action to the result of the query action.
>   Now, if the execution of a statement within a `do-clause` fails with an error, it will be propagated to the nearest enclosing failure-handling statement.

Reviewer: ```suggestion
- Fixed a bug that resulted in errors returned from a `do` clause of a query action not being propagated correctly to  the result of the query action. Now, if the execution of a statement within a `do` clause fails with an error, it will be propagated to the nearest enclosing failure-handling statement.
```

Author: The idea changed with your above suggestion. Hope this is fine.
```suggestion
- Fixed a bug that incorrectly propagated errors returned from a `do` clause of a query action to the result of the query action. Now, if the execution of a statement within a `do` clause fails with an error, it will be propagated to the nearest enclosing failure-handling statement.
```

Reviewer: Ah yes, misunderstood. LGTM.

```ballerina
public function main() {
    error? queryActResult = from int i in 1 ... 3
        do {
            check validateAndGetError(); // error: invalid usage of the 'check' expression operator; 
                                         // no matching error return type(s) in the enclosing invokable
        };
}

function validateAndGetError() returns error? {
    return error("Invalid error");
}
```

> - Fixed a bug where the use of `on-fail` led to uninitialized variables at runtime. 
>   Now, the compiler would emit errors for possible uninitialized variables.

Reviewer: ```suggestion
- Fixed a bug that resulted in variables that may or may not be initialized in an `on fail` clause not being identified as potentially uninitialized variables.
```
